spec/controllers/users_controller_spec.rb

```diff
@@ -0,0 +1,186 @@
+        expect(response).to render_template(:new)
+      end
+    end
+
+    context "when account already has a user" do
+      before do
+        sign_in account_with_user
+        get :new
+      end
+
+      it "redirects to root path" do
+        expect(response).to redirect_to(root_path)
+      end
+
+      it "sets a flash message" do
+        expect( flash[:warning]).to eq(I18n.t("noti.account_already_has_user"))
+      end
+    end
+
+    context "when account is not present" do
+      before do
+        get :new
+      end
+
+      it "redirects to new account registration path" do
+        expect(response).to redirect_to(new_account_registration_path)
+      end
+    end
+  end
+
+  describe "POST #create" do
+    before do
+      sign_in account_without_user
+    end
+
+    context "with valid user params" do
+      it "creates a new user" do
+        expect {
+          post :create, params:{user: user_params}
+        }.to change(User, :count).by(1)
+      end
+
+      it "redirects to root path" do
+        post :create, params:{user: user_params}
+        expect(response).to redirect_to(root_path)
+      end
+    end
+
+    context "with invalid user params" do
+      before do
+        user_params[:citizen_id] = nil
+      end
+
+      it "does not create a new user" do
+        expect {
+          post :create, params:{user: user_params}
+        }.not_to change(User, :count)
+      end
+
+      it "renders the new template again" do
+        post :create, params:{user: user_params}
+        expect(response).to redirect_to(new_user_path)
+      end
+    end
+  end
+
+  describe "GET #show" do
+    before do
+      sign_in account_with_user
+    end
+    context "when user exists" do
+      it "renders the show template" do
+        get :show, params:{id: account_with_user.user.id}
+        expect(response).to render_template(:show)
+      end
+    end
+    context "when user does not exist" do
+      before do
+        get :show, params:{id: -1}
+      end
+      it "redirects to the root path with a flash message" do
+        expect(response).to redirect_to(root_path)
+      end
+      it "sets a flash message" do
+        expect(flash[:danger]).to eq(I18n.t("noti.user_not_found"))
+      end
+    end
+    
+  end
+
+  describe "GET #edit" do
+    before do
+      sign_in account_with_user
+    end
+    context "when user exists" do
+      it "renders the edit template" do
+        get :edit, params:{id: account_with_user.user.id}
+        expect(response).to render_template(:edit)
+      end
+    end
+
+    context "when user does not exist" do
+      before do
+        get :edit, params:{id: -1}
+      end
+      it "redirects to the root path with a flash message" do
+        expect(response).to redirect_to(root_path)
+      end
+      it "sets a flash message" do
+        expect(flash[:danger]).to eq(I18n.t("noti.user_not_found"))
+      end
+    end
+  end
+
+  describe "PATCH #update" do
```

Comment:
describe, context, it sẽ dùng trong những trường hợp nào

Reply:
describe: Tạo ra một example group để nhóm các các test liên quan đến một chủ đề nhất định lại với nhau.
context: Là để mô tả các điều kiện/ ngữ cảnh mà test case sẽ chạy.
it: được sử dụng để xác định một test case cụ thể. Mỗi it thường đại diện cho một test đơn lẻ.